common/flatpak-transaction-log.c

```diff
@@ -0,0 +1,287 @@
+                                      guint         prop_id,
+                                      const GValue *value,
+                                      GParamSpec   *pspec)
+{
+  FlatpakTransactionLog *self = FLATPAK_TRANSACTION_LOG (object);
+
+  switch (prop_id)
+    {
+    case PROP_PATH:
+      g_assert (self->path == NULL);
+      self->path = g_value_dup_object (value);
+      break;
+
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+    }
+}
+
+/* For efficiency's sake, we have a special "append only" log format here. This
+ * is a subset of JSON, but is not valid JSON in its own right - rather, each
+ * line is a new JSON object. Applications wishing to parse this file should
+ * read and parse each line independently. */
+static gboolean
+append_node_to_log_file (GFile        *file,
+                         JsonNode     *node,
+                         GCancellable *cancellable,
+                         GError       **error)
+{
+  g_autoptr(GFile) parent = g_file_get_parent (file);
+  g_autoptr(GFileOutputStream) stream = NULL;
+  g_autofree gchar *serialized = NULL;
+  gsize serialized_len = 0;
+  gsize written = 0;
+  g_autoptr(JsonGenerator) generator = NULL;
+  g_autoptr(GError) my_error = NULL;
+
+  if (!g_file_make_directory_with_parents (parent, cancellable, &my_error))
+    {
+      if (!g_error_matches (my_error, G_IO_ERROR, G_IO_ERROR_EXISTS))
+        {
+          g_propagate_error (error, g_steal_pointer (&my_error));
+          return FALSE;
+        }
+    }
+
+  generator = json_generator_new ();
+  json_generator_set_root (generator, node);
+  json_generator_set_pretty (generator, FALSE);
+
+  /* Resize the buffer to accomadate one extra newline and write that
+   * in place at serialized_len */
+  serialized = json_generator_to_data (generator, &serialized_len);
+  serialized = g_realloc (serialized, (serialized_len + 1) * sizeof (gchar));
+  serialized[serialized_len++] = '\n';
+  serialized[serialized_len] = '\0';
+
+  stream = g_file_append_to (file,
+                             G_FILE_CREATE_NONE,
+                             cancellable,
+                             error);
+
+  return g_output_stream_write_all (G_OUTPUT_STREAM (stream),
+                                    (gconstpointer) serialized,
+                                    serialized_len,
+                                    &written,
+                                    cancellable,
+                                    error);
+}
+
+#define EVENT_LOG_VERSION "1"
+
+static JsonNode *
+build_base_event (const gchar *event)
+{
+  g_autoptr(JsonBuilder) builder = json_builder_new ();
+
+  json_builder_begin_object (builder);
+
+  json_builder_set_member_name (builder, "event");
+  json_builder_add_string_value (builder, event);
+
+  json_builder_set_member_name (builder, "version");
+  json_builder_add_string_value (builder, EVENT_LOG_VERSION);
+  json_builder_end_object (builder);
```

Comment:
I'm a bit uncertain about the version field. I mean, how would you use it? If we want to just add stuff to the json object that is fine without a version field. Consumers can just check for the new fields. A version is just useful if we're for some reason doing a completely backwards incompatible changes so that consumers of the existing versions have to ignore the new nodes. I think we should avoid such changes, and in the unlikely case we need them, just go with a different filename for the log.

Reply:
Okay - I guess if we only care about forward compatibility I can drop the field.